Source/CarthageKit/Project.swift

```diff
@@ -675,19 +681,46 @@ public final class Project {
 				return self.buildOrderForResolvedCartfile(resolvedCartfile, dependenciesToInclude: dependenciesToBuild)
 			}
 			.flatMap(.Concat) { dependency -> SignalProducer<BuildSchemeProducer, CarthageError> in
-				let dependencyPath = self.directoryURL.URLByAppendingPathComponent(dependency.project.relativePath, isDirectory: true).path!
+				let project = dependency.project
+				let version = dependency.version.commitish
+
+				let dependencyPath = self.directoryURL.URLByAppendingPathComponent(project.relativePath, isDirectory: true).path!
 				if !NSFileManager.defaultManager().fileExistsAtPath(dependencyPath) {
 					return .empty
 				}
 
-				return buildDependencyProject(dependency.project, self.directoryURL, withOptions: options, sdkFilter: sdkFilter)
+				var options = options
+				let cleanDerivedDataIfNeeded: SignalProducer<(), CarthageError>
+
+				if options.derivedDataPath != nil {
+					cleanDerivedDataIfNeeded = .empty
+				} else {
+					let derivedDataPerDependency = CarthageDependencyDerivedDataURL
+						.URLByAppendingPathComponent(self.directoryURL.lastPathComponent!, isDirectory: true)
+						.URLByAppendingPathComponent(project.name, isDirectory: true)
+					let derivedDataVersioned = derivedDataPerDependency.URLByAppendingPathComponent(version, isDirectory: true)
+					options.derivedDataPath = derivedDataVersioned.URLByResolvingSymlinksInPath?.path
+
+					let fileManager = NSFileManager.defaultManager()
+					cleanDerivedDataIfNeeded = fileManager
+						.carthage_enumeratorAtURL(derivedDataPerDependency, includingPropertiesForKeys: [], options: [ .SkipsSubdirectoryDescendants ], catchErrors: true)
+						.flatMap(.Concat) { _, URL -> SignalProducer<(), CarthageError> in
+							if URL != derivedDataVersioned {
+								_ = try? fileManager.removeItemAtURL(URL)
+							}
+							return .empty
+						}
```

Comment:
If there is an existing derived data dir for the specific commitish of the dependency, and we are trying to build the dependency with a different commitish, we should remove (clean) the directory and do a clean build instead of adding `clean` action to `xcodebuild` command.